README.md
# Paymob Accept Python Library

The Accept Python library provides convenient access to the Accept API from applications written in the Python language.
It includes a pre-defined set of classes for API resources that initialize themselves dynamically from API responses
which makes it compatible with a wide range of versions of the Accept API.

## Installation

```sh
pip install accept-paymob
```

Install from source with:

```sh
python setup.py install
```

### Requirements

- Python 3.4+

## Usage

The library needs to be configured with your account's secret key which is available in
your [Accept Dashboard](https://accept.paymob.com/portal2/en/settings).

Set `api_key` to its value:

```python

from accept.payment import *

API_KEY = "<API-KEY>"

accept = AcceptAPI(API_KEY)

# Authentication Request
auth_token = accept.retrieve_auth_token()
print(auth_token)

# Order Registration
OrderData = {
    "auth_token": auth_token,
    "delivery_needed": "false",
    "amount_cents": "1100",
    "currency": "EGP",
    "merchant_order_id": 125,  # UNIQUE
    "items": [
        {
            "name": "ASC1515",
            "amount_cents": "500000",
            "description": "Smart Watch",
            "quantity": "1"
        },
        {
            "name": "ERT6565",
            "amount_cents": "200000",
            "description": "Power Bank",
            "quantity": "1"
        }
    ],
    "shipping_data": {
        "apartment": "803",
        "email": "claudette09@exa.com",
        "floor": "42",
        "first_name": "Clifford",
        "street": "Ethan Land",
        "building": "8028",
        "phone_number": "+86(8)9135210487",
        "postal_code": "01898",
        "extra_description": "8 Ram , 128 Giga",
        "city": "Jaskolskiburgh",
        "country": "CR",
        "last_name": "Nicolas",
        "state": "Utah"
    },
    "shipping_details": {
        "notes": " test",
        "number_of_packages": 1,
        "weight": 10,
        "weight_unit": "Kilogram",
        "length": 100,
        "width": 100,
        "height": 100,
        "contents": "product of some sorts"
    }
}
order = accept.order_registration(OrderData)
print(order)

# Payment Key Request
Request = {
    "auth_token": auth_token,
    "amount_cents": "1500",
    "expiration": 3600,
    "order_id": order.get("id"),
    "billing_data": {
        "apartment": "803",
        "email": "claudette09@exa.com",
        "floor": "42",
        "first_name": "Clifford",
        "street": "Ethan Land",
        "building": "8028",
        "phone_number": "+86(8)9135210487",
        "shipping_method": "PKG",
        "postal_code": "01898",
        "city": "Jaskolskiburgh",
        "country": "CR",
        "last_name": "Nicolas",
        "state": "Utah"
    },
    "currency": "EGP",
    "integration_id": 246701,  # https://accept.paymob.com/portal2/en/PaymentIntegrations
    "lock_order_when_paid": "false"
}
payment_token = accept.payment_key_request(Request)
print(payment_token)

# Payments API [Kiosk, Mobile Wallets , Cash, Pay With Saved Token]
identifier = "cash"
payment_method = "CASH"
transaction = accept.pay(identifier, payment_method, payment_token)
print(transaction)

# Auth-Capture Payments
transaction00 = accept.capture_transaction(transaction_id="7608793", amount_cents=1000)
print(transaction00)

# Refund Transaction
transaction01 = accept.refund_transaction(transaction_id="7608793", amount_cents=10)
print(transaction01)

# Void Transaction
transaction02 = accept.void_transaction(transaction_id="7608793")
print(transaction02)

# Retrieve Transaction
transaction03 = accept.retrieve_transaction(transaction_id="7608793")
print(transaction03)

# Inquire Transaction
transaction_inquire = accept.inquire_transaction(merchant_order_id="123", order_id="10883471")
print(transaction_inquire)

# Tracking
order_10883471_track = accept.tracking(order_id="10883471")
print(order_10883471_track)

# Preparing Package
# This will return a pdf file url to be printed.
package = accept.preparing_package(order_id="10883471")  
print(package)

# IFrame URL
iframeURL = accept.retrieve_iframe(iframe_id="230796", payment_token=payment_token)
print(iframeURL)

# Loyalty Checkout
response = accept.loyalty_checkout(transaction_reference='', otp='123', payment_token=payment_token)
print(response)

```



CHANGELOG.md
# Changelog

## 1.0.0 - 2021-05-01

* Start ...


tests/__init__.py
"""
Paymob Accept Testing.
"""

tests/test_payment.py
from random import *

from accept.payment import *

API_KEY = "<API-KEY>"
accept = AcceptAPI(API_KEY)


def test_retrieve_auth_token():
    auth_token = accept.retrieve_auth_token()
    assert len(auth_token) > 100


def test_order_registration():
    OrderData = {
        "auth_token": accept.retrieve_auth_token(),
        "delivery_needed": "false",
        "amount_cents": "1100",
        "currency": "EGP",
        "merchant_order_id": randint(1, 1000),
        "items": [
            {
                "name": "ASC1515",
                "amount_cents": "500000",
                "description": "Smart Watch",
                "quantity": "1"
            },
            {
                "name": "ERT6565",
                "amount_cents": "200000",
                "description": "Power Bank",
                "quantity": "1"
            }
        ],
        "shipping_data": {
            "apartment": "803",
            "email": "claudette09@exa.com",
            "floor": "42",
            "first_name": "Clifford",
            "street": "Ethan Land",
            "building": "8028",
            "phone_number": "+86(8)9135210487",
            "postal_code": "01898",
            "extra_description": "8 Ram , 128 Giga",
            "city": "Jaskolskiburgh",
            "country": "CR",
            "last_name": "Nicolas",
            "state": "Utah"
        },
        "shipping_details": {
            "notes": " test",
            "number_of_packages": 1,
            "weight": 10,
            "weight_unit": "Kilogram",
            "length": 100,
            "width": 100,
            "height": 100,
            "contents": "product of some sorts"
        }
    }
    order = accept.order_registration(OrderData)

    assert order.get("amount_cents") == 1100


def test_payment_key_request():
    OrderData = {
        "auth_token": accept.retrieve_auth_token(),
        "delivery_needed": "false",
        "amount_cents": "1100",
        "currency": "EGP",
        "merchant_order_id": randint(1, 1000),
        "items": [
            {
                "name": "ASC1515",
                "amount_cents": "500000",
                "description": "Smart Watch",
                "quantity": "1"
            },
            {
                "name": "ERT6565",
                "amount_cents": "200000",
                "description": "Power Bank",
                "quantity": "1"
            }
        ],
        "shipping_data": {
            "apartment": "803",
            "email": "claudette09@exa.com",
            "floor": "42",
            "first_name": "Clifford",
            "street": "Ethan Land",
            "building": "8028",
            "phone_number": "+86(8)9135210487",
            "postal_code": "01898",
            "extra_description": "8 Ram , 128 Giga",
            "city": "Jaskolskiburgh",
            "country": "CR",
            "last_name": "Nicolas",
            "state": "Utah"
        },
        "shipping_details": {
            "notes": " test",
            "number_of_packages": 1,
            "weight": 10,
            "weight_unit": "Kilogram",
            "length": 100,
            "width": 100,
            "height": 100,
            "contents": "product of some sorts"
        }
    }
    order = accept.order_registration(OrderData)

    Request = {
        "auth_token": accept.retrieve_auth_token(),
        "amount_cents": "1500",
        "expiration": 3600,
        "order_id": order.get("id"),
        "billing_data": {
            "apartment": "803",
            "email": "claudette09@exa.com",
            "floor": "42",
            "first_name": "Clifford",
            "street": "Ethan Land",
            "building": "8028",
            "phone_number": "+86(8)9135210487",
            "shipping_method": "PKG",
            "postal_code": "01898",
            "city": "Jaskolskiburgh",
            "country": "CR",
            "last_name": "Nicolas",
            "state": "Utah"
        },
        "currency": "EGP",
        "integration_id": 246701,  # https://accept.paymob.com/portal2/en/PaymentIntegrations
        "lock_order_when_paid": "false"
    }

    payment_token = accept.payment_key_request(Request)
    assert len(payment_token) > 100


accept/payment.py
import requests

from accept.endpoints import *


class AcceptAPI:
    def __init__(self, API_KEY):
        self.api_key = API_KEY
        self.url = URLs()

    def retrieve_auth_token(self):
        """
        Authentication Request:
        :return: token: Authentication token, which is valid for one hour from the creation time.
        """
        data = {'api_key': self.api_key}
        r = requests.post(self.url.authentication_url(), json=data)
        token = r.json().get('token')
        return token

    def order_registration(self, data):
        """
        Order Registration API
        :param data: Order data
        :return: registered order
        """
        r = requests.post(self.url.order_registration_url(), json=data)
        order = r.json()
        return order

    def payment_key_request(self, data):
        r = requests.post(self.url.payment_key_url(), json=data)
        payment_token = r.json().get('token')
        return payment_token

    def pay(self, identifier, payment_method, payment_token):
        data = {
            "source": {
                "identifier": identifier,
                "subtype": payment_method
            },
            "payment_token": payment_token
        }
        r = requests.post(self.url.payment_url(), json=data)
        transaction = r.json()
        return transaction

    def capture_transaction(self, transaction_id, amount_cents):
        auth_token = self.retrieve_auth_token()
        data = {
            "auth_token": auth_token,
            "transaction_id": transaction_id,
            "amount_cents": amount_cents
        }
        r = requests.post(self.url.capture_transaction_url(), json=data)

        transaction = r.json()
        return transaction

    def refund_transaction(self, transaction_id, amount_cents):
        auth_token = self.retrieve_auth_token()
        data = {
            "auth_token": auth_token,
            "transaction_id": transaction_id,
            "amount_cents": amount_cents
        }
        r = requests.post(self.url.refund_transaction_url(), json=data)

        transaction = r.json()
        return transaction

    def void_transaction(self, transaction_id):
        auth_token = self.retrieve_auth_token()
        data = {
            "transaction_id": transaction_id,
        }
        r = requests.post(self.url.void_transaction_url().format(token=str(auth_token)), json=data)

        transaction = r.json()
        return transaction

    def retrieve_transaction(self, transaction_id):
        auth_token = self.retrieve_auth_token()
        r = requests.get(self.url.retrieve_transaction_url().format(id=str(transaction_id),
                                                                    token=str(auth_token)))
        transaction = r.json()
        return transaction

    def inquire_transaction(self, merchant_order_id, order_id):
        auth_token = self.retrieve_auth_token()
        payload = {
            "auth_token": auth_token,
            "merchant_order_id": merchant_order_id,
            "order_id": order_id
        }

        r = requests.post(self.url.inquire_transaction_with_order_details_url(), json=payload)

        transaction = r.json()
        return transaction

    def tracking(self, order_id):
        auth_token = self.retrieve_auth_token()
        r = requests.get(self.url.tracking_url().format(order_id=str(order_id),
                                                        token=str(auth_token)))
        order = r.json()
        return order

    def preparing_package(self, order_id):
        auth_token = self.retrieve_auth_token()
        pdf_link = self.url.preparing_package_url().format(order_id=str(order_id), token=str(auth_token))
        return pdf_link

    def loyalty_checkout(self, transaction_reference, otp, payment_token):
        data = {
            "reference": transaction_reference,
            "otp": otp,
            "payment_token": payment_token
        }

        r = requests.post(self.url.loyalty_checkout_url(), json=data)
        response = r.json()
        return response

    def retrieve_iframe(self, iframe_id, payment_token):
        iframe_url = self.url.iframe_url().format(iframe_id=iframe_id, payment_token=payment_token)
        return iframe_url


accept/endpoints.py
class URLs:
    def __init__(self):
        self.base_url = "https://accept.paymob.com/api/"

        # auth
        self.auth = "auth/tokens"

        # ecommerce
        self.order = "ecommerce/orders"
        self.inquire_transaction = "ecommerce/orders/transaction_inquiry"
        self.tracking = "ecommerce/orders/{order_id}/delivery_status?token={token}".format(order_id="{order_id}",
                                                                                           token="{token}")
        self.preparing_package = "ecommerce/orders/{order_id}/airway_bill?token={token}".format(order_id="{order_id}",
                                                                                                token="{token}")

        # acceptance
        self.payment_key = "acceptance/payment_keys"
        self.payment = "acceptance/payments/pay"
        self.capture = "acceptance/capture"
        self.refund = "acceptance/void_refund/refund"
        self.void = "acceptance/void_refund/void?token={token}".format(token="{token}")
        self.retrieve_transaction = "acceptance/transactions/{id}?token={token}".format(id="{id}", token="{token}")
        self.loyalty_checkout = "acceptance/loyalty_checkout"
        self.iframe = "acceptance/iframes/{iframe_id}?payment_token={payment_token}".format(iframe_id="{iframe_id}",
                                                                                            payment_token="{payment_token}")

    def base_url(self):
        return self.base_url

    def authentication_url(self):
        return self.base_url + self.auth

    def order_registration_url(self):
        return self.base_url + self.order

    def payment_key_url(self):
        return self.base_url + self.payment_key

    def payment_url(self):
        return self.base_url + self.payment

    def capture_transaction_url(self):
        return self.base_url + self.capture

    def refund_transaction_url(self):
        return self.base_url + self.refund

    def void_transaction_url(self):
        return self.base_url + self.void

    def retrieve_transaction_url(self):
        return self.base_url + self.retrieve_transaction

    def inquire_transaction_with_order_details_url(self):
        return self.base_url + self.inquire_transaction

    def tracking_url(self):
        return self.base_url + self.tracking

    def preparing_package_url(self):
        return self.base_url + self.preparing_package

    def loyalty_checkout_url(self):
        return self.base_url + self.loyalty_checkout

    def iframe_url(self):
        return self.base_url + self.iframe


examples/Usage.py
from accept.payment import *

# These values are from Paymob.accept API.
API_KEY = "ZXlKMGVYQWlPaUpLVjFRaUxDSmhiR2NpT2lKSVV6VXhNaUo5LmV5SnVZVzFsSWpvaWFXNXBkR2xoYkNJc0ltTnNZWE56SWpvaVRXVnlZMmhoYm5RaUxDSndjbTltYVd4bFgzQnJJam96TkRNemZRLkh0NS1VclIxS0t1N2MzVkIxbllMRGlYNUpTSy1jTWlKS1Vod1ExNnFGcmV2US05T3lvWDJUTnQyUm5saTZsMUZFblVVLVVPOThzS3p1MjgwVE92aXBn"

if __name__ == "__main__":
    accept = AcceptAPI(API_KEY)

    # Authentication Request
    print("-- Authentication Request --")

    auth_token = accept.retrieve_auth_token()
    print(auth_token)

    #################################################################################

    # Order Registration
    print("-- Order Registration --")

    OrderData = {
        "auth_token": auth_token,
        "delivery_needed": "false",
        "amount_cents": "1100",
        "currency": "EGP",
        "merchant_order_id": 125,
        "items": [
            {
                "name": "ASC1515",
                "amount_cents": "500000",
                "description": "Smart Watch",
                "quantity": "1"
            },
            {
                "name": "ERT6565",
                "amount_cents": "200000",
                "description": "Power Bank",
                "quantity": "1"
            }
        ],
        "shipping_data": {
            "apartment": "803",
            "email": "claudette09@exa.com",
            "floor": "42",
            "first_name": "Clifford",
            "street": "Ethan Land",
            "building": "8028",
            "phone_number": "+86(8)9135210487",
            "postal_code": "01898",
            "extra_description": "8 Ram , 128 Giga",
            "city": "Jaskolskiburgh",
            "country": "CR",
            "last_name": "Nicolas",
            "state": "Utah"
        },
        "shipping_details": {
            "notes": " test",
            "number_of_packages": 1,
            "weight": 10,
            "weight_unit": "Kilogram",
            "length": 100,
            "width": 100,
            "height": 100,
            "contents": "product of some sorts"
        }
    }

    order = accept.order_registration(OrderData)
    print(order)

    #################################################################################

    # Payment Key Request
    print("-- Payment Key Request --")

    Request = {
        "auth_token": auth_token,
        "amount_cents": "1500",
        "expiration": 3600,
        "order_id": order.get("id"),
        "billing_data": {
            "apartment": "803",
            "email": "claudette09@exa.com",
            "floor": "42",
            "first_name": "Clifford",
            "street": "Ethan Land",
            "building": "8028",
            "phone_number": "+86(8)9135210487",
            "shipping_method": "PKG",
            "postal_code": "01898",
            "city": "Jaskolskiburgh",
            "country": "CR",
            "last_name": "Nicolas",
            "state": "Utah"
        },
        "currency": "EGP",
        "integration_id": 246701,  # https://accept.paymob.com/portal2/en/PaymentIntegrations
        "lock_order_when_paid": "false"
    }

    payment_token = accept.payment_key_request(Request)
    print(payment_token)

    #################################################################################

    # Payments API [Kiosk, Mobile Wallets , Cash, Pay With Saved Token]
    print("-- Payments API --")

    identifier = "cash"
    payment_method = "CASH"

    transaction = accept.pay(identifier, payment_method, payment_token)

    print(transaction)

    #################################################################################

    # Auth-Capture Payments
    print("-- Auth-Capture Payments --")

    transaction00 = accept.capture_transaction("7608793", 1000)
    print(transaction00)

    #################################################################################

    # Refund Transaction
    print("-- Refund Transaction --")

    transaction01 = accept.refund_transaction("7608793", 10)
    print(transaction01)

    #################################################################################

    # Void Transaction
    print("-- Void Transaction --")

    transaction02 = accept.void_transaction("7608793")
    print(transaction02)

    #################################################################################

    # Retrieve Transaction
    print("-- Retrieve Transaction --")

    transaction03 = accept.retrieve_transaction("7608793")
    print(transaction03)

    #################################################################################

    # Inquire Transaction
    print("-- Inquire Transaction --")

    transaction_inquire = accept.inquire_transaction("123", "10883471")
    print(transaction_inquire)

    #################################################################################

    # Tracking
    print("-- Tracking --")

    order_10883471_track = accept.tracking("10883471")
    print(order_10883471_track)

    #################################################################################
    # This will return a pdf file to be printed.
    # Preparing Package
    print("-- Preparing Package --")

    package = accept.preparing_package("10883471")
    print(package)

    #################################################################################

    # IFrame URL
    print("-- IFrame URL --")

    iframeURL = accept.retrieve_iframe("230796", payment_token)
    print(iframeURL)
